src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,596 @@
+        #endif
+        hasCompleted = !stopRequested;
+        isRunning = false;
+    }
+}
+
+/* ---------- CCoinsSelectionBranchAndBound ---------- */
+
+/* ---------- CCoinsSelectionForNotes ---------- */
+
+CCoinsSelectionForNotes::CCoinsSelectionForNotes(std::vector<std::pair<CAmount, size_t>> _amountsAndSizes,
+                                                 CAmount _targetAmount,
+                                                 CAmount _targetAmountPlusOffset,
+                                                 size_t _availableTotalSize,
+                                                 std::vector<CAmount> _joinsplitsOutputsAmounts)
+                                                 : CCoinsSelectionAlgorithmBase(CoinsSelectionAlgorithmType::FOR_NOTES,
+                                                                                _amountsAndSizes,
+                                                                                _targetAmount,
+                                                                                _targetAmountPlusOffset,
+                                                                                _availableTotalSize),
+                                                                                numberOfJoinsplitsOutputsAmounts {(unsigned int)_joinsplitsOutputsAmounts.size()},
+                                                                                joinsplitsOutputsAmounts(_joinsplitsOutputsAmounts)
+{
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    iterations = 0;
+    #endif
+}
+
+CCoinsSelectionForNotes::~CCoinsSelectionForNotes()
+{
+}
+
+bool CCoinsSelectionForNotes::Reset()
+{
+    bool resetActuallyDone = CCoinsSelectionAlgorithmBase::Reset();
+
+    if (resetActuallyDone)
+    {
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        iterations = 0;
+        #endif
+    }
+
+    return resetActuallyDone;
+}
+
+void CCoinsSelectionForNotes::Solve()
+{
+    if (!isRunning)
+    {
+        Reset();
+
+        isRunning = true;
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        uint64_t microsecondsBefore = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+        #endif
+
+        int windowBackIndex = maxIndex;
+        int windowFrontIndex = maxIndex;
+        bool admissibleFound = false;
+        bool bestAdmissibleFound = false; // "best" for this specific algorithm implementation
+        bool breakOuterLoop = false;
+
+        for (; !stopRequested && !breakOuterLoop && windowBackIndex >= 0; --windowBackIndex)
+        {
+            // quick reset before restarting with sliding window back index decreased by one position
+            for (int index = 0; index < problemDimension; ++index)
+            {
+                tempSelection[index] = 0;
+            }
```

Comment:
As you've changed from array to vector you have plenty of options to make the code more concise, more idiomatic ... instead of using this verbose loops with random accessor

Option 1
`tmpSelection.assign(problemDimension, 0);`

Option 2
`memset(&tmpSelection[0], 0, problemDimension * sizeof(tmpSelection[0]));`

Option 3 (which is the best describing the intent)
`std::fill(tmpSelection.begin(), tmpSelection.end(), 0)`

Reply:
I agree with your suggestion.
Modified accordingly.